staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go

```diff
@@ -180,34 +180,49 @@ func (ts *azureTokenSource) Token() (*azureToken, error) {
 	defer ts.lock.Unlock()
 
 	var err error
+	// 1. use cache if available
 	token := ts.cache.getToken(azureTokenKey)
+
+	// 2. load persist if no cache
 	if token == nil {
 		token, err = ts.retrieveTokenFromCfg()
-		if err != nil {
-			token, err = ts.source.Token()
-			if err != nil {
-				return nil, fmt.Errorf("acquiring a new fresh token: %v", err)
-			}
-		}
-		if !token.token.IsExpired() {
+
+		// set cfg to cache 
+		if err == nil {
+			klog.V(4).Infof("Saving cache after load cfg")
 			ts.cache.setToken(azureTokenKey, token)
-			err = ts.storeTokenInCfg(token)
-			if err != nil {
-				return nil, fmt.Errorf("storing the token in configuration: %v", err)
-			}
 		}
 	}
-	if token.token.IsExpired() {
+	
+	// quick exit for caching scenario
+	if token != nil && !token.token.IsExpired() {
+		return token, nil
+	}
+
+	// 3. extend if valid token but expired
+	if token != nil && token.token.IsExpired() {
+		klog.V(4).Infof("Refreshing token.")
 		token, err = ts.refreshToken(token)
 		if err != nil {
-			return nil, fmt.Errorf("refreshing the expired token: %v", err)
+			klog.V(4).Infof("Failed to refresh expired token: %v", err)
 		}
-		ts.cache.setToken(azureTokenKey, token)
-		err = ts.storeTokenInCfg(token)
+	}
+
+	// 4. real-auth if invalid token
+	if token == nil {
+		token, err = ts.source.Token()
 		if err != nil {
-			return nil, fmt.Errorf("storing the refreshed token in configuration: %v", err)
+			return nil, fmt.Errorf("acquiring a new fresh token: %v", err)
```

Comment:
nit: `Failed to acquire a new token: %v`

Reply:
The original code was always using "<step>: %v", err for err message, maybe I can refactor it in this PR?